variable/include/scipp/variable/transform.h

```diff
@@ -98,6 +99,54 @@ template <int N, class T> static constexpr auto get(const T &index) noexcept {
 
 } // namespace iter
 
+template <size_t N_Operands>
+inline constexpr auto stride_special_cases =
+    std::array<std::array<scipp::index, N_Operands>, 0>{};
+
+template <>
+inline constexpr auto stride_special_cases<1> =
+    std::array<std::array<scipp::index, 1>, 2>{{{1}, {0}}};
+
+template <>
+inline constexpr auto stride_special_cases<2> =
+    std::array<std::array<scipp::index, 2>, 4>{
+        {{1, 1}, {0, 1}, {1, 0}, {0, 0}}};
```

Comment:
We will need different special cases for in-place an non-inplace. In particular, for non-inplace we generally have operations such as `out = a + b`. We know that `out` always has stride 1, and we'd like optimization, e.g., for `{1,1,1}, {1,0,1}, {1,1,0}, ...` (there the first `1` is fixed, so this 3-arg case is actually two args). Unless I miss something it is currently not possible to specify this independently?

Reply:
No, it's not currently possible. I was trying to avoid duplicating code.

One approach might be: List a whole lot of special cases globally and provide a constexpr filter function to dispatch. This could select relevant strides based on in-place / non-in-place or even other criteria (op?). But I'm not sure how to avoid stamping out the whole set of `dispatch_inner_loop` every time and hope for the compiler to remove the unused cases.

Alternatively, we might be able to pass the strides cases array as a template parameter somehow. But we are a bit limited in C==17 in that respect...

Interestingly, the benchmarks are still really good even though there are no special cases for 3 operands.

Reply:
The problem is also that as written now we are compiling a number of cases that are unreachable (all those starting with `0` in the non-inplace case). Since we already struggle with compile times this should be avoided.

Can't you just template `stride_special_case` (or make a helper function to get it)?  For example in:
```cpp
 } else if constexpr (!in_place &&
                       detail::stride_special_cases<N_Operands>[I][0] > 1) {
```
use `stride_special_case<in_place>`?

Reply:
Tried with adding in_place as template param to `stride_special_cases` and using those for non-in-place:
```.cpp
template <>
inline constexpr auto stride_special_cases<2, false> =
    std::array<std::array<scipp::index, 2>, 1>{{{1, 1}}};

template <>
inline constexpr auto stride_special_cases<3, false> =
    std::array<std::array<scipp::index, 3>, 3>{
        {{1, 1, 1}, {1, 0, 1}, {1, 1, 0}}};
```

Benchmarked the build with gcc v10.2.0 on an Intel(R) Core(TM) i7-9850H with 12 threads. Configured with `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON -DCMAKE_BUILD_TYPE=RELEASE`.
It took _slightly_ longer to compile compared to the code proposed in this PR: 20m21s (new) vs 19m49s (old). The binary size gives a similar picture: 16MiB (new) vs 15Mib (old).

I also tried with 
```.cpp
inline constexpr auto stride_special_cases<3> =
    std::array<std::array<scipp::index, 3>, 3>{
        {{1, 1, 1}, {1, 0, 1}, {1, 1, 0}}};
```
added to the code that doesn't discriminate between in-place and non-in-place. And the results are 19m45s and 16MiB.

Note that all three cases produce different optimisations and will likely lead to different run times. So the numbers here are not easily comparable. However, since there does not seem to be any significant difference between those cases, I would prefer sticking with the simpler code.

Reply:
I consider it an *requirement* that different special cases can be specified for in-place and non-in-place. There are some specialized (internal) applications requiring 3 or 4 arguments, i.e., the lists of cases will grow longer than they are now. We do not want to generate those cases for both in-place and non-inplace.

Furthermore, all cases that have the stride=0 for the first param are redundant for non-in-place.

I don't see where the code gets complicated? We simply have two cases instead of one. One could even argue that it becomes simpler to understand, since then special cases map to specific applications, whereas with the combined handling a case like `{1, 1}` is used for `a += b` and `sin(x)`, which is maybe confusing?

Finally, since we are shipping this as a library, it seems like an odd approach to include dead code branches, making the library larger than necessary.

Reply:
I agree with your last point.
But adding separate stride cases for (non-)in-place does increase complexity. Not much, but still. The only reason for having the `in_place` template param in the first place is to select `call` or `call_in_place`. Unless I'm overlooking something, those two could be unified which would remove the need for the template param. At that point, I would rather have a more general way of selecting strides than just in-place or not. This would allow us to provide optimised implementations for specific, important operations without adding them to all of them. However, I am not sure how that can be done yet.

Reply:
We have much bigger complexity issues elsewhere. This feels tiny (unless I misunderstand what is required for the change), and has subtantial downsides.